File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest_sum = float('inf')
        min_diff = float('inf')

        for i in range(len(nums) - 2):
            left, right = i + 1, len(nums) - 1

            while left < right:
                current_sum = nums[i] + nums[left] + nums[right]
                current_diff = abs(current_sum - target)

                if current_diff < min_diff:
                    min_diff = current_diff
                    closest_sum = current_sum

                if current_sum < target:
                    left += 1
                else:
                    right -= 1

        return closest_sum        

File: 0016-3sum-closest/0016-3sum-closest.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
    n = nums.length
    nums.sort!
    closest_sum = nums[0] + nums[1] + nums[2]
    min_diff = (target - closest_sum).abs

    (0...n - 2).each do |i|
        left = i + 1
        right = n - 1

        while left < right
            sum = nums[i] + nums[left] + nums[right]
            diff = (target - sum).abs

            if diff < min_diff
                min_diff = diff
                closest_sum = sum
            end

            if sum < target
                left += 1
            elsif sum > target
                right -= 1
            else
                return target # The sum is equal to the target, so it's the closest.
            end
        end
    end

    closest_sum
end
